libraries/plugins/elasticsearch/elasticsearch_plugin.cpp

```diff
@@ -0,0 +1,373 @@
+   bulks.block_data = bs;
+   bulks.additional_data = vs;
+
+   std::string alltogether = fc::json::to_string(bulks);
+
+   // bulk header before each line, op_type = create to avoid dups, index id will be ath id(2.9.X).
+   std::string _id = fc::json::to_string(ath.id);
+   bulk.push_back("{ \"index\" : { \"_index\" : \"graphene\", \"_type\" : \"data\", \"op_type\" : \"create\", \"_id\" : "+_id+" } }"); // header
+   bulk.push_back(alltogether);
+}
+
+void elasticsearch_plugin_impl::sendBulk(std::string _elasticsearch_node_url, bool _elasticsearch_logs)
+{
+
+   // curl buffers to read
+   std::string readBuffer;
+   std::string readBuffer_logs;
+
+   std::string bulking = "";
+
+   bulking = boost::algorithm::join(bulk, "\n");
+   bulking = bulking + "\n";
+   bulk.clear();
+
+   //wlog((bulking));
+
+   struct curl_slist *headers = NULL;
+   curl_slist_append(headers, "Content-Type: application/json");
+   std::string url = _elasticsearch_node_url + "_bulk";
+   curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+   curl_easy_setopt(curl, CURLOPT_POST, true);
+   curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+   curl_easy_setopt(curl, CURLOPT_POSTFIELDS, bulking.c_str());
+   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+   curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&readBuffer);
+   curl_easy_setopt(curl, CURLOPT_USERAGENT, "libcrp/0.1");
+   //curl_easy_setopt(curl, CURLOPT_VERBOSE, true);
+   curl_easy_perform(curl);
+
+   long http_code = 0;
+   curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code);
+   if(http_code == 200) {
+      // all good, do nothing
+   }
+   else if(http_code == 429) {
+      // repeat request?
+   }
+   else {
+      // exit everything ?
+   }
+
+   if(_elasticsearch_logs) {
+      auto logs = readBuffer;
+      // do logs
+      std::string url_logs = _elasticsearch_node_url + "logs/data/";
+      curl_easy_setopt(curl, CURLOPT_URL, url_logs.c_str());
+      curl_easy_setopt(curl, CURLOPT_POST, true);
+      curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDS, logs.c_str());
+      curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+      curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *) &readBuffer_logs);
+      curl_easy_setopt(curl, CURLOPT_USERAGENT, "libcrp/0.1");
+      //curl_easy_setopt(curl, CURLOPT_VERBOSE, true);
+      //ilog("log here curl: ${output}", ("output", readBuffer_logs));
+      curl_easy_perform(curl);
+
+      http_code = 0;
+      curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code);
+      if(http_code == 200) {
+         // all good, do nothing
+      }
+      else if(http_code == 429) {
+         // repeat request?
+      }
+      else {
+         // exit everything ?
+      }
+   }
+}
+
+} // end namespace detail
+
+elasticsearch_plugin::elasticsearch_plugin() :
+   my( new detail::elasticsearch_plugin_impl(*this) )
+{
+   curl = curl_easy_init();
```

Comment:
if `curl` will be used in several plugins, need a `NULL` check here, so it won't get initialized for more than one time. 

Actually it's better to move it and related variables/functions into a dedicated utility class, and use a singleton model. So when other code need it, no need to include this plugin.

Reply:
if another plugin use curl it it will definitely use another instance not the same one. the same instance/session is used to elasticsearch work only.